Library/OcAppleKernelLib/MkextContext.c

```diff
@@ -0,0 +1,1040 @@
+  UINT32                CurrentOffset;
+  UINT32                NewOffset;
+  UINT32                PlistOffset;
+  UINT32                PlistCompSize;
+  UINT32                PlistFullSize;
+  UINT32                BinOffset;
+  UINT32                BinCompSize;
+  UINT32                BinFullSize;
+
+  UINT8                 *PlistBuffer;
+  XML_DOCUMENT          *PlistXml;
+  XML_NODE              *PlistBundles;
+  UINT32                PlistBundlesCount;
+  XML_NODE              *PlistBundle;
+  UINT32                PlistBundleIndex;
+  UINT32                PlistBundleCount;
+  CONST CHAR8           *PlistBundleKey;
+  XML_NODE              *BundleExecutable;
+
+  MKEXT_V2_FILE_ENTRY   *MkextExecutableEntry;
+  MKEXT_V2_FILE_ENTRY   *MkextOutExecutableEntry;
+  CHAR8                 *BinaryOffsetStrings;
+  UINT32                BinaryOffsetStringsSize;
+
+  ASSERT (Buffer != NULL);
+  ASSERT (BufferSize > 0);
+
+  MkextDecompress = OutBufferSize > 0;
+  if (MkextDecompress) {
+    ASSERT (OutBuffer != NULL);
+    ASSERT (OutMkextSize != NULL);
+  }
+
+  if (BufferSize < sizeof (MKEXT_CORE_HEADER)
+    || !OC_TYPE_ALIGNED (MKEXT_CORE_HEADER, Buffer)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  MkextHeader   = (MKEXT_HEADER_ANY *) Buffer;
+  MkextSize     = SwapBytes32 (MkextHeader->Common.Length);
+  MkextVersion  = SwapBytes32 (MkextHeader->Common.Version);
+  NumKexts      = SwapBytes32 (MkextHeader->Common.NumKexts);
+
+  if (MkextHeader->Common.Magic != MKEXT_INVERT_MAGIC
+    || MkextHeader->Common.Signature != MKEXT_INVERT_SIGNATURE
+    || BufferSize != MkextSize
+    || OcOverflowAddU32 (NumKexts, NumReservedKexts, &NumMaxKexts)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Mkext v1.
+  //
+  if (MkextVersion == MKEXT_VERSION_V1) {
+    if (OcOverflowMulAddU32 (sizeof (MKEXT_V1_KEXT), NumMaxKexts, sizeof (MKEXT_V1_HEADER), &CurrentOffset)) {
+      return EFI_INVALID_PARAMETER;
+    }
+    CurrentOffset = MKEXT_ALIGN (CurrentOffset);
+
+    if (MkextDecompress) {
+      //
+      // Copy header.
+      //
+      CopyMem (OutBuffer, Buffer, sizeof (MKEXT_V1_HEADER));
+      MkextHeaderOut = (MKEXT_HEADER_ANY *) OutBuffer;
+    }
+
+    //
+    // Process plists and binaries if present, decompressing as needed.
+    //
+    for (Index = 0; Index < NumKexts; Index++) {
+      PlistFullSize   = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.FullSize);
+      BinFullSize     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.FullSize);
+
+      if (MkextDecompress) {
+        //
+        // Decompress entry.
+        //
+        PlistOffset     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.Offset);
+        PlistCompSize   = SwapBytes32 (MkextHeader->V1.Kexts[Index].Plist.CompressedSize);
+        BinOffset       = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.Offset);
+        BinCompSize     = SwapBytes32 (MkextHeader->V1.Kexts[Index].Binary.CompressedSize);
+
+        if (CurrentOffset > OutBufferSize) {
```

Comment:
This can never be `> OutBufferSize`. At most it should be `==`. Inverse and make this an assertion.

In addition to an assertion there should be two checks:
- whether MKEXT_ALIGN (PlistFullSize) and MKEXT_ALIGN (BinFullSize) are >= than the original values
- whether adding MKEXT_ALIGN (PlistFullSize) and MKEXT_ALIGN (BinFullSize) lead to CurrentOffset results in an overflow OR `> OutBufferSize`.